File: 665.non-decreasing-array.ts
/*
 * @lc app=leetcode id=665 lang=typescript
 *
 * [665] Non-decreasing Array
 */

// @lc code=start
function checkPossibility(nums: number[]): boolean {
  let errors = 0;

  for (let i = 0; i < nums.length - 1; i++) {
    if (nums[i] > nums[i + 1]) {
      errors++;

      if (errors > 1) {
        return false;
      }

      if (i > 0 && nums[i - 1] > nums[i + 1]) {
        nums[i + 1] = nums[i];
      } else {
        nums[i] = nums[i + 1];
      }
    }
  }

  return true;
}
// @lc code=end

File: python/665.non-decreasing-array.py
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        if len(nums) <= 2:
            return True
        flag = False
        for i in range(1, len(nums)):
            if nums[i] < nums[i - 1]:
                if flag:
                    return False
                flag = True
                if i > 1 and nums[i] < nums[i - 2]:
                    nums[i] = nums[i - 1]
        return True